test-data/unit/check-classes.test

```diff
@@ -2065,3 +2065,167 @@ class A: pass
 class B(object, A): # E: Cannot determine consistent method resolution order (MRO) for "B"
     def readlines(self): pass
     __iter__ = readlines
+
+[case testSelfTypeInstance]
+from typing import TypeVar
+
+T = TypeVar('T')
+
+class A:
+    def copy(self: T) -> T: ...
+
+class B(A):
+    pass
+
+reveal_type(A().copy)  # E: Revealed type is 'def () -> __main__.A'
+reveal_type(B().copy)  # E: Revealed type is 'def () -> __main__.B'
+reveal_type(A().copy())  # E: Revealed type is '__main__.A'
+reveal_type(B().copy())  # E: Revealed type is '__main__.B'
+# FIX: Why partial type?
+# reveal_type(A.copy(A()))
+# reveal_type(B.copy(B()))
+
+
+[case testSelfTypeReturn]
+from typing import TypeVar, Type
+
+T = TypeVar('T')
+
+def _type(self: T) -> Type[T]: ...
+
+class A:
+    def copy(self: T) -> T:
+        if B():
+            return A()
+        elif A():
+            return B()
+        return _type(self)()
+
+class B(A):
+    pass
+
+[out]
+main: note: In member "copy" of class "A":
+main:10: error: Incompatible return value type (got "A", expected "T")
+main:12: error: Incompatible return value type (got "B", expected "T")
+
+[case testSelfTypeClass]
+from typing import TypeVar, Type
+
+T = TypeVar('T')
+
+class A:
+    @classmethod
+    def new(cls: Type[T]) -> T:
+        return cls()
```

Comment:
A few additional things to test:

* `reveal_type(cls)`
* `reveal_type(cls())`
* `cls(2)` (invalid arguments)